From c570569080f71c14bfb75aea7469b143805863a3 Mon Sep 17 00:00:00 2001
From: Arianna Avanzini <avanzini.arianna@gmail.com>
Date: Thu, 17 Jan 2013 13:11:32 +0100
Subject: [PATCH] block, bfq: do not update peak rate if no I/O was done
 during the service round

It may happen that, due to I/O errors, the active queue is expired
without any of its requests being actually completed. If this happens,
do not update the peak rate.
---
 block/bfq-cgroup.c  |    2 +-
 block/bfq-iosched.c |    3 +++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/block/bfq-cgroup.c b/block/bfq-cgroup.c
index e8efb03..49d74a0 100644
--- a/block/bfq-cgroup.c
+++ b/block/bfq-cgroup.c
@@ -694,7 +694,7 @@ static struct cgroup_subsys_state *bfqio_create(struct cgroup *cgroup)
 }
 
 /*
- * We cannot support shared io contexts, as we have no mean to support
+ * We cannot support shared io contexts, as we have no means to support
  * two tasks with the same ioc in two different groups without major rework
  * of the main bic/bfqq data structures.  By now we allow a task to change
  * its cgroup only if it's the only owner of its ioc; the drawback of this
diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 1bbd21b..5ea128c 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -1430,6 +1430,9 @@ static int bfq_update_peak_rate(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 	ktime_t delta;
 	int update = 0;
 
+	if (bfqq->entity.service == 0)
+		return 0;
+
 	if (!bfq_bfqq_sync(bfqq) || bfq_bfqq_budget_new(bfqq))
 		return 0;
 
-- 
1.7.10.4

